backend/timeline/service.go

```diff
@@ -40,37 +41,70 @@ type service struct {
 	config Config
 	lock   sync.RWMutex
 	nextID int
-	events []*timelinepb.Event
+	db     *memdb.MemDB
 
 	notifier *channels.Notifier
 }
 
+// envelope is a wrapper around an event for indexing.
+type envelope struct {
+	Event *timelinepb.Event
+	ID    uint64
+}
+
 var _ timelineconnect.TimelineServiceHandler = (*service)(nil)
 
 func Start(ctx context.Context, config Config) error {
 	config.SetDefaults()
 
 	logger := log.FromContext(ctx).Scope("timeline")
 	ctx = log.ContextWithLogger(ctx, logger)
-	svc := &service{
-		config:   config,
-		events:   make([]*timelinepb.Event, 0),
-		nextID:   0,
-		notifier: channels.NewNotifier(ctx),
+	svc, err := newService(ctx, config)
+	if err != nil {
+		return fmt.Errorf("failed to create timeline service: %w", err)
 	}
 
 	go svc.reapCallEvents(ctx)
 
 	logger.Debugf("Timeline service listening on: %s", config.Bind)
-	err := rpc.Serve(ctx, config.Bind,
+	if err := rpc.Serve(ctx, config.Bind,
 		rpc.GRPC(timelineconnect.NewTimelineServiceHandler, svc),
-	)
-	if err != nil {
+	); err != nil {
 		return fmt.Errorf("timeline service stopped serving: %w", err)
 	}
 	return nil
 }
 
+func newService(ctx context.Context, config Config) (*service, error) {
+	schema := &memdb.DBSchema{
```

Comment:
I can see pulling in an in-memory DB being worthwhile if we had relatively complex indexing requirements, but given our current data model of an append only array this seems a bit excessive?

I suspect we could get _very_ similar performance by just:

1. Sorting `.events` by ID to ensure the array is completely sorted (if it's possible IDs could arrive out of order)
2. Using `sort.Search()` to do a binary search.

Reply:
Ids are allocated on the server, so I don't think it is possible for them to be out of order, it's only timestamps that are not guaranteed.

Reply:
Yep, we probably want to allow sorting with timestamps in the future as well, as well as possibly by modules. In which case a slice won't work anymore.

Reply:
Decided we will not need secondary indices, and went with just searching based on event id.